### ISSUE TITLE
Routing in Angular Based Web Components using @angular/elements

### ISSUE THREAD

**Issue by tuckerjt07** [angular/angular#34377]
# 🚀 feature request

### Relevant Package

 This feature request is for @angular/elements and @angular/router

### Description
 When attempting to use an Angular app to host a second Angular app wrapped as a Web Component routing does not work for the hosted app.

### Describe the solution you'd like
 It would be nice if an instance of the hosting app could be passed in so that the router.forChild() could be used to register the routes for the web component. Another option  would be to rework both elements and router so that multiple forRoot instances of the router do not cause issues.

### Describe alternatives you've considered
 I have tried named outlets and ui-router. UI-router got me the closest by using the Angular Router in the hosting application and ui-router in the hosted component. The issue I ran into there was that I could not directly navigate by URL into a sub-route using ui-router.

**Comment by tuckerjt07** [comment]
I've managed to get close to what I want using the steps in this article, https://medium.com/@timon.grassl/how-to-use-routing-in-angular-web-components-c6a76449cdb . The issue is that using the named router-outlets is causing other routing issues, particularly around navigating between two primary routes to a named outlet in the second route and also with redirectUrls that contain named outlets.

**Comment by tuckerjt07** [comment]
Another update, here is a gist of where I'm at with this, https://gist.github.com/tuckerjt07/8c8c7fd55208c4c23f8adb0750e15ee8 . It's actually working pretty close to how I want it to. Now the only issue is I am getting some extraneous states in the browser's nav stack.

**Comment by BufferUnderrunOK** [comment]
Even with Timon Grassl's code, we still have multiple problems with deep linking to items belonging to the child applications. We really need a way for Angular applications to register their own routes with the Parent applications' Router. Think of it as the inverse of `loadChildren`.

**Comment by tuckerjt07** [comment]
Agreed, the biggest issue I am getting now is that when my component's router is used it removes any of the history that was there before from the nav stack.

**Comment by BufferUnderrunOK** [comment]
But let's just leave this as completely unimportant and move on with things nobody asked for. 👍

**Comment by konstantintieber** [comment]
My colleague @fboeller recently published a npm package for exactly this use case: https://github.com/fboeller/ngx-elements-router

We are using this approach for developing Angular microfrontends in web-components today and I'm sure he'll continue to update that package as we find new bugs/missing behaviour.

**Comment by lilletech** [comment]
I had the problem. I've removed lazy loading and used routermodule.forroot to make routing working

**Comment by rotty3000** [comment]
It seems fundamentally broken to not have good isolation within the routing engine. I'm not an angular expert nor even a JavaScript expert, but the mere fact that the routing engine in Angular does not apparently have any isolation after this much time is sad. I wrongly assumed that such a mature framework would just "have a solution" for this. It's not even about inheritance from one engine to another.

If an _instance_ of a routing engine could set a base when it is instantiated and if many routing engines could set their own bases separately then creating a hierarchy of engines should flow from a parent setting the handler of one if it's routes to any other instance, and also support the case for peers from any other technology (i.e. other web components from other frameworks.)

My (potentially super naive) 2 cents.

**Comment by bbech** [comment]
+1

**Comment by arjunasuresh3** [comment]
+1

**Comment by marrej** [comment]
+1

**Comment by rotty3000** [comment]
+1

**Comment by SchnWalter** [comment]
Guys, please stop with the `+1`. On the right sidebar, you have a "Subscribe" button, please use that.

**Comment by VIDIZMO-bilal-ahmed** [comment]
Is there any progress on this issue?

**Comment by jainank** [comment]
👍

**Comment by jainank** [comment]
FYI --
To include the community in the feature request process, we open voting for 60 days. Anyone can cast a vote for the request with a thumbs-up (👍) reaction. When a feature request reaches 20 or more upvotes, we formally consider the feature request. Alternatively, the bot closes the request.

**Comment by marceloluzbmo** [comment]
👍

**Comment by Lindsor** [comment]
👍

**Comment by BufferUnderrunOK** [comment]
👍

**Comment by a29340** [comment]
👍🏻

**Comment by rotty3000** [comment]
:+1:

**Comment by TheLangdaTyagi** [comment]
👍

**Comment by bala3011s** [comment]
👍

**Comment by konstantintieber** [comment]
[Fabian Böller](https://github.com/fboeller) and me just published a video that covers everything about how we ship web component-based micro frontends with Angular.

Here's the two video chapters that cover our integration of the Angular router of the surrounding application with the web component:
- Propagate route changes to micro frontend: https://www.youtube.com/watch?v=ee17YczpCpU&t=2373s
- Propagate route changes to platform: https://www.youtube.com/watch?v=ee17YczpCpU&t=5271s

Link to the project on GitHub: https://github.com/fboeller/microfrontends-with-angular/tree/recording

I hope this can help to inspire the discussions on how to support this in the Angular router.

**Comment by blidblid** [comment]
If your web components are small, you could handle routing natively. It's super simplistic, but it just works.

<script src="https://gist.github.com/blidblid/b920440c7e6f68bd24a2257dca489be5.js"></script>

To create components, use `*ngComponentOutlet` over `router-outlet`.

**Comment by bh3605** [comment]
Crazy this hasn't been solved yet. I'm using two standalone components in a monorepo set up and I'm sharing the same route-config between the two of them because I'm using angular-architect/mod-fed to load the 2nd stand alone component as a web component and I'm able to get the route to trigger, but the context.outlet object is null and therefore won't load the 2nd component.

root app
```
const routeProvider = provideRouter(routes, withDebugTracing())
setGlobalRoutes(routeProvider);

export const appConfig: ApplicationConfig = {
  providers: [
    routeProvider,
    // {provide: LocationStrategy, useClass: MultiLocationStrategy},
    {
      provide: ENVIRONMENT,
      useFactory: resolveEnvironment
    },
    importProvidersFrom(HttpClientModule, ModuleFederationToolsModule)
  ]
};
```

root app html
```
<mft-wc-wrapper [options]="item"></mft-wc-wrapper>
```

root app component
```
export class MyRootComponent {
  item: WebComponentWrapperOptions = {
    type: "manifest",
    remoteName: 'claims',
    exposedModule: 'RecentClaims',
    // elementName: "claims-recent-claims"
    elementName: 'recent-claims'
  };
}
```

root app routes
```
export const routes: Routes = [
    {
    path: 'claims',
    loadComponent: () => loadRemoteModule({
      type: "manifest",
      remoteName: "claims",
      exposedModule: "AppComponent"
    }).then(m => m.AppComponent)
  }
]
```

shared lib containing the setGlobalRoutes function
```
export const routerSubject = new BehaviorSubject<EnvironmentProviders>(null as unknown as EnvironmentProviders);

export function setGlobalRoutes(router: EnvironmentProviders) {
  routerSubject.next(router);
}
```

imported 2nd mfe exposed as a web component so it can be dynamically loaded. RecentClaims has a routerLink to /claims and the router _almost_ succeeds in doing the navigation, but atlas it does not.
```
@Component({
  selector: 'claims-recent-claims',
  standalone: true,
  imports: [CommonModule, RouterLink],
  templateUrl: './recent-claims.component.html',
  styleUrls: ['./recent-claims.component.css'],
  schemas: [CUSTOM_ELEMENTS_SCHEMA]
})
export class RecentClaimsComponent {
  constructor(private claimsSvc: ClaimsService) {}
}

// createEnvironmentInjector

const app = await createApplication(
  {
    providers: [
      {
        provide: ENVIRONMENT,
        useFactory: resolveEnvironment
      },
      routerSubject.getValue()
    ]
  }
);

const recentClaims = createCustomElement(RecentClaimsComponent, {
  injector: app.injector
});
customElements.define('recent-claims', recentClaims);
```

**Comment by mauriziocescon** [comment]
Hey! 

We have the same problem in a huge enterprise app where having one single build (no matter how much optimised) was impossible: too many devs, too many teams, too much code, too many deadlines, ... say: the usual conditions where a micro frontend architecture makes sense. 

In the end, we decided to build a custom solution (lib of components / tools / style encapsulation included) where the routing part is managed like this: https://github.com/mauriziocescon/myws (just sharing for helping, not promoting it 😅).

The setup in the repo is pretty standard: 

- a `host` (shell) loading the various parts of the application,
- 3 "sections" (`mf1`, `mf2`, `mf3`) available at the urls:
  - `mf1` => /mf1,
  - `mf2` => /mf2,
  - `mf3` => /mf3,
- 1 "standalone" (`mf4`) not attached to any url, 

Since it's ng-cli based, all the apps are v18... but it should work fine with different ng versions as well (other frameworks are not considered). 

In the repo there are 2 `app.routes.ts` files, once for "local-development" (lazy loading) and another one for "deployment". Each app bundles everything, angular included (in our real case we use the usual stuff like CDN, ... to compensate). 

In the example, the creation of components is mostly done programmatically (rather than using templates). Therefore there's no need for `CUSTOM_ELEMENTS_SCHEMA`... on the other hand, `globalThis` is used to shared routing information between the `host` and `mf`. 

I hope it might help somebody having to deal with the same problem. 

Just my 2 cents on the topic!


PS (@atscott): First of all: thanks a lot for the amazing work on the router in the last few years! Second: I perfectly understand this is not an easy thing to solve at angular level (if even possible)... but I also see the topic popping up pretty often. Maybe it might make sense to try to support at least a subset of (not extreme) cases around some common agreement driven by an initial ng-team proposal and a RFC.
